File: src/generator.c
#include <stdio.h>
#include <stdlib.h>
#include "labyrinth.h"
#include "generator.h"

/*
char** generate_labyrinth(){
    srand( time( NULL ) );

    char** labyrinth = init_labyrinth();
    

    for (int i=0; i <21;i++){
        for (int j=0; j <22; j++){
            labyrinth[i][j] = "|";
            
        }
    }

    
    int x = rand()%22;
    int y = rand()%23;
    char* current_cell = labyrinth[x][y];
    
    labyrinth[x][y] = "";
    char * frontiers = get_frontiers(labyrinth, x, y);
    int n = sizeof(frontiers)/sizeof(frontiers[0]);
    
    while(n != 0){
        x = rand()%22;
        y = rand()%23;
        //frontiers[n+1] = labyrinth[x][y];
        
        //Connect the chosen frontier and a random neighbor.
        connect_to_neighbor(labyrinth,x,y);

        //Add the frontiers of the chosen frontier to the frontier list.
        tmp = get_frontiers(labyrinth,x,y);
        int size = sizeof(tmp)/sizeof(tmp[0]);

        for ( i=0; i < size; i++ ){
            frontiers[n+(1+i)] = tmp[i];
        }

    }

    return labyrinth;
}

char* get_frontiers(char** labyrinth, int x, int y){

    char* frontiers = malloc(30* sizeof(char));
    int tmp[] = {(0,2),(2,0),(2,2),(0,-2),(-2,0),(-2,-2)};
    
    for (int i = 0; i<6; i++){
        if (x+tmp[i][0] < 22 && y+tmp[i][1] < 21){
            if ( labyrinth[x+tmp[i][0]][y+tmp[i][1]] =="|"){
                frontiers[i] = 
            } 
        }
    }
    return frontiers;
}

void save_labyrinth(char** labyrinth){

    char* filename = "3.txt";

    // open the file for writing
    FILE *fp = fopen(filename, "w");
    if (fp == NULL)
    {
        printf("Error opening the file %s", filename);
        return -1;
    }


    // write to the text file
    for (int i = 0; i < 21; i++){
        fprintf(fp, "\n");

        for(int j = 0; j<22; j++){
            fprintf(fp, "%s", labyrinth[i][j] );
        }
    }
        

    // close the file
    fclose(fp);

    return 0;
}

int main(){

    save_labyrinth(generate_labyrinth());
}

*/

File: generator/generator.py
import random

n = 75
m = 75

def generate_labyrinth():
    labyrinth = []

    for i in range(n):
        labyrinth.append([])
        for j in range(m):
            labyrinth[i].append('|')
    
    x = random.randint(0,n)
    y = random.randint(0,m)
    labyrinth[x][y] = ' '
    frontiers = get_frontiers(labyrinth,x,y)
    
    
    
    while len(frontiers) != 0:

        print("frontiers list :", frontiers)
        #Choose a random frontier (ie. its coordinates)
        k = random.randint(0, len(frontiers)-1 )
        (fx, fy) = frontiers[k]
        frontiers.pop(k)

        #Connect the chosen frontier and a random neighbor.
        neighbors = get_neighbors(labyrinth, fx, fy)
        if neighbors:
            nx , ny = neighbors[random.randint(0, len(neighbors)-1)]
            labyrinth[(fx + nx) //2][(fy + ny)//2] = ' '
            labyrinth[fx][fy] = ' '
        
        #Add the frontiers of the chosen frontier to the frontier list.
        new_frontiers = get_frontiers(labyrinth, fx, fy)
        for f in new_frontiers:
            if f not in frontiers:
                frontiers.append(f)


    for i in range(n):
        if labyrinth[i][0] == ' ':
            labyrinth[i][0] = '|'
        if labyrinth[i][-1] == ' ':
            labyrinth[i][-1] = '|'  
    for j in range(m):
        if labyrinth[0][j] == ' ':
            labyrinth[0][j] = '|'
        if labyrinth[-1][j] == ' ':
            labyrinth[-1][j] = '|'  
            
    return labyrinth


def get_frontiers(labyrinth, x, y):

    frontiers = []
    tmp = [(0,2),(2,0),(0,-2),(-2,0)]

    for i in range(len(tmp)):
        a = tmp[i][0]
        b = tmp[i][1]

        if (0 <= x + a < n) and (0 <= y + b < m) :
            if labyrinth[x+a][y+b] == '|':
                frontiers.append((x + a , y + b) )

    return frontiers


def get_neighbors(labyrinth, x, y):
    neighbors = []
    tmp = [(0,2),(2,0),(0,-2),(-2,0)]

    for i in range(len(tmp)):
        a = tmp[i][0]
        b = tmp[i][1]

        if (0 <= x + a < n) and (0 <= y + b < m):
            
            if labyrinth[x+a][y+b] == ' ':
                neighbors.append((x + a, y + b) )
    
    return neighbors


def save_labyrinth(labyrinth, i):
    with open("../levels/"+str(i)+".txt", "a") as f:
        f.truncate(0)

        for i in range(len(labyrinth)):
            
            for j in range(len(labyrinth[i])):

                f.write(labyrinth[i][j])
            f.write("\n")

if __name__ == "__main__":

    for i in range(0, 6):

        save_labyrinth(generate_labyrinth(),i)